diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt

```diff
@@ -56,15 +56,26 @@ class DataClassesRule(configRules: List<RulesConfig>) : DiktatRule("data-classes
     @Suppress("UnsafeCallOnNullableType", "FUNCTION_BOOLEAN_PREFIX", "ComplexMethod")
     private fun ASTNode.canBeDataClass(): Boolean {
         val isNotPropertyInClassBody = findChildByType(CLASS_BODY)?.let { (it.psi as KtClassBody).properties.isEmpty() } ?: true
+        val constructorParametersNames = mutableListOf<String>()
         val hasPropertyInConstructor = findChildByType(PRIMARY_CONSTRUCTOR)
             ?.let { constructor ->
                 (constructor.psi as KtPrimaryConstructor)
                     .valueParameters
-                    .run { isNotEmpty() && all { it.hasValOrVar() } }
+                    .run { isNotEmpty() && all { if (it.hasValOrVar()) true else { constructorParametersNames.add(it.name!!); false } } }
             } ?: false
         if (isNotPropertyInClassBody && !hasPropertyInConstructor) {
             return false
         }
+        // if parameter of the primary constructor is used in init block then it is hard to refactor this class to data class
+        if (constructorParametersNames.isNotEmpty()) {
+            val initBlocks = findChildByType(CLASS_BODY)?.getAllChildrenWithType(CLASS_INITIALIZER)
+            initBlocks?.forEach { init ->
+                val refExpressions = init.findAllNodesWithSpecificType(REFERENCE_EXPRESSION)
+                if (refExpressions.any { it.text !in constructorParametersNames }) {
```

Comment:
Is it right? Shouldn't it be 'any in'? Also please add tests with more parameters and more complex init blocks

Reply:
Should be. Added more tests